포스트(해시태그) 관련 API 테스트 작성 및 구현

dlawnsdh opened the pull request:
# 작성한 파일
- MVC
  - PostController
  - PostRepository
    - PostRepositoryCustom
    - PostRepositoryCustomImpl
  - HashtagRepository
  - PostHashtagRepository
  - LikesRepository
  - PostService
    - StorageService
    - FileSystemStorageService
 - DTO
   - UserAccountDto
   - PostRequest
   - PostResponse
 - Exception
   - PostException
   - PostExceptionType
 - Test
   - PostControllerTest
   - PostServiceTest
# 변경된 파일
- UserAccountDto 재추가
# 삭제된 파일
- 해당 없음
# 변경된 사항
- `UserAccountDto` 재추가 -> `UserPrincipal` 로 대체
- `Post` Controller, Service 성공 테스트 작성
- `Post` Controller, Service, Repository 작성
- `Hashtag`, `PostHashtag`, `Likes` Repository 작성
- Request , Response DTO 작성
- `Post` 비지니스 로직에서 발생할 수 있는 예외 정의
## 관련 이슈
- This closes #20

dlawnsdh committed:
#20 - UserAccountDto 다시 추가

포스트 컨트롤러에서 서비스 단으로 유저 정보를 넘겨줄 때
사용하기 위해 재 추가, 추후 `UserPrincipal`로 대체

dlawnsdh committed:
#20 - PostRequest, PostResponse DTO 작성

포스트 생성, 수정을 위한 dto와 사용자에게 필요한 정보를 
담은 dto 추가

PostResponse는 댓글 정보도 추가되어야 함

dlawnsdh committed:
#20 - Post 비지니스 로직 예외 정의

dlawnsdh committed:
#20 - Post 관련 레포지토리 작성

querydsl을 사용하여 필요한 쿼리문 구현

dlawnsdh committed:
#20 - 기타 레포지토리 생성

`Hashtag`, `PostHashtag`, `Likes` 레포지토리 생성

dlawnsdh committed:
#20 - 로컬 스토리지 서비스 구현

추후 S3 버킷으로 저장소 옮길 예정

dlawnsdh committed:
#20 - PostController 작성

포스트 관련 요청을 처리하는 컨트롤러 작성

좋아요 관련 기능 작성 필요

dlawnsdh committed:
#20 - PostService 작성

포스트 관련 비지니스 로직을 처리

좋아요 관련 로직 구현 필요

dlawnsdh committed:
#20 - PostController 슬라이스 테스트 작성

 예외를 처리하는 테스트 추가 필요

dlawnsdh committed:
#20 - PostService 슬라이스 테스트 작성

 예외를 처리하는 테스트 추가 필요

sGOM commented on simple-community/src/main/java/com/project/simplecommunity/service/PostService.java:
```diff
@@ -0,0 +1,143 @@
+            if (postRequest.title() != null) post.setTitle(postRequest.title());
+            if (postRequest.content() != null) post.setContent(postRequest.content());
+            if (postRequest.hashtags() != null) {
+                postHashtagRepository.deleteAllByPost_Id(postId);
```
`deleteAll`은 내부적으로 반복문을 순회하며 `delete`를 한번 씩 호출하는 거라서 해당 포스트에 달린 해쉬태그만큼 Query문이 날아가게 되어있음

내 생각에는 `deleteAll`보다는 `deleteAllInBatch`를 한번 알아보고 사용하는게 나을거 같어
[deleteAll vs deleteAllInBatch](https://stackoverflow.com/questions/58231638/whats-the-difference-between-deleteallinbatch-and-deleteall)

dlawnsdh replied:
조언 굳!!!!!! 👍 
알아보고 고쳐볼게

dlawnsdh replied:
https://velog.io/@hssarah/JPA-deleteAll-vs-deleteAllInBatch-vs-Query-%EC%82%AC%EC%9A%A9
해당 글 보고 @Query 사용해서 하려고

sGOM commented on simple-community/src/main/java/com/project/simplecommunity/repository/querydsl/PostRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,57 @@
+        QPostHashtag postHashtag = QPostHashtag.postHashtag;
+
+        JPQLQuery<Post> query = from(post)
+                .innerJoin(post.postHashtags, postHashtag)
```
`one-to-many`관계에서 `one`쪽을 기준으로 join을 하게 되면 반환되는 테이블이 페이징하고자 하는 대상의 레코드보다 많아지기 때문에 전부 메모리에 올린 후에 선택하는 방식으로 작동하기 때문에 성능 면에서도 문제가 있고 OOM문제도 발생할 수 있어서

혹시 해당 Query를 실행했는데 다음과 같은 문구가 나오면 내가 언급한 문제가 발생하는거야
`HHH000104: firstResult/maxResults specified with collection fetch; applying in memory!`

만약에 이런 문제가 발생한다면 join하기 보다는 양방향 매핑을 통해서 해결하는게 더 좋을거 같아
[ JPA fetch join + Paging (limit) 처리 시 발생하는 문제](https://hungseong.tistory.com/68)

dlawnsdh replied:
내가 원하는건 해시태그로 해당 포스트를 찾아내는 건데 from(post) 이거는 필요한게 아닌건가?
지금 매핑은 Post(`OneToMany`) <-> (`ManyToOne`)PostHashtag(`ManyToOne`) <-> (`OneToMany`)Hahstag 
양방향으로 이렇게 되어있어

처음에는 ` .innerJoin(post.postHashtags, postHashtag)` 이거 다음에 이너조인을 하나 더 넣었었는데
`.where(hashtag.hashtagName.in(hashtagNames))` 이렇게 해주니까 `PostHashtag` 랑 `Hashtag` 를 알아서 조인하더라고

그래서 지금 쿼리문보면
select
        p1_0.id,
        p1_0.content,
        p1_0.created_at,
        p1_0.created_by,
        p1_0.modified_at,
        p1_0.modified_by,
        p1_0.title,
        p1_0.total_like_count,
        p1_0.user_id 
    from
        post p1_0 
    join
        post_hashtag p2_0 
            on p1_0.id=p2_0.post_id 
    join
        hashtag h1_0 
            on h1_0.id=p2_0.hashtag_id 
    where
        h1_0.hashtag_name in(?,?) 
    order by
        p1_0.created_at desc limit ?,
        ?
이렇게 날아감 조인 두개가 성능 많이 잡아먹긴 해
일단 더 찾아볼게

dlawnsdh replied:
일단 문제는 발생안하는 것으로 확인

sGOM replied:
맞네 fetch join 을 안 하니까 문제가 발생하지 않는구나
`PostHashtag`는 fetch join 안 했으니까 Lazy Loading하고

resposne mapping 할 때 `PostHashtag`랑 join column 설정 해둔 `Hashtag` 한 번 select해서 가져오는구나
이 방식이면 페이지네이션도 정상적으로 되고 쿼리문도 그리 많지 않아서 문제 없을듯

sGOM commented on simple-community/src/main/java/com/project/simplecommunity/service/PostService.java:
```diff
@@ -0,0 +1,143 @@
+                postHashtagRepository.deleteAllByPost_Id(postId);
+                postRepository.flush();
+
+                for (String hashtagName : postRequest.hashtags()) {
```
이 부분 반복문을 순회하면서 `save`를 호출하는건 한번 호출마다 새로운 트랜잭션이 생성되기 때문에 `saveAll`을 통해서 한번의 트랜잭션으로 해결하는 것보다는 오버헤드가 발생할 수 있는데 둘 다 구현을 확인해보면 Query의 발생 횟수는 같으니까 내 생각에는 크게 차이는 없을 것 같긴 함

아래에 링크 한 번 읽어보고 판단해보는게 나을 듯
[save vs saveAll](https://www.baeldung.com/spring-data-save-saveall)

dlawnsdh replied:
해시태그 특성상 엄청 많은 데이터가 한번에 저장되는게 아니긴한데 하나의 트랜잭션에서 수행되는 `saveAll` 로 수정할게

sGOM reviewed (commented):
Spring Data JPA를 사용한 부분만 몇 개 댓글 달았으니 확인해줘

다른 부분은 다 괜찮은 것 같음 👍

dlawnsdh committed:
#20 - PostHashtag: delete 수정

기존 방식은 삭제 한 건 당 하나의 쿼리가 수행되었는데
@Query 를 사용하여 직접 쿼리문을 작성해
In 방식으로 한번의 쿼리에 삭제되도록 수정

dlawnsdh committed:
#20 - PostRepositoryCustomImpl: 해당 이슈 주석 작성

HHH000104: firstResult/maxResults specified with collection fetch; applying in memory!

해당 경고 문구가 나오지 않는 이유 주석으로 작성

dlawnsdh committed:
#20 - PostService : saveAll() 사용

하나의 트랜잭션 안에서  수행되는 saveAll 사용

sGOM reviewed (approved):
굿 👍  수정하느라 수고했어 ㅎ

근데 1개 conflict 난다 ㅋㅋㅋ

dlawnsdh committed:
#20 - UserAccountDto 제거

충돌이 발생된 UserAccountDto 삭제하고 
그에 따른 코드 변경 사항 적용 

1. UserAccountDto -> UserPrincipal 로 적용
 - PostController, PostService
 - PostControllerTest, PostServiceTest

2. PostService 로직 변경에 따른 테스트 코드 @Disabled 적용
 - save() -> saveAll() 변경됨, 후에 테스트 코드 반영 필요

sGOM reviewed (approved):
수고했으
